feat(stepfunctions): add support for EncryptionConfiguration (#30959)
aws/aws-cdk: packages/aws-cdk-lib/aws-stepfunctions/README.md

});
```

### Creating a StateMachine with Encryption using a Customer Managed Key

> ```

Reviewer: Probably could use an accompanying description

Author: Will add, thanks

});
```

> ### Creating a StateMachine with Encryption using a Customer Managed Key

Author: Ok fair, will update.

```
const kmsKey = new kms.Key(stack, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
      stateMachineName: 'StateMachineWithCMKEncryptionConfiguration',
      definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'Pass'))),
      stateMachineType: sfn.StateMachineType.STANDARD,
      kmsKey: kmsKey,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(60)
    });
```

### Creating a StateMachine with CWL Encryption using a Customer Managed Key,
You can encrypt data sent to CloudWatch Logs. To use encrypted logging, you must set `enableEncryptedLogging` to `true` and provide the `logs?` prop.
```
const stateMachineKmsKey = new kms.Key(this, 'StateMachineKey');
const logGroupKmsKey = new kms.Key(this, 'LogGroupKmsKey');

/**
     * We need to grant the service principal encrypt and decrypt permissions since passing
     *  a KMS key when creating a LogGroup doesn't automatically grant the service principal encrypt/decrypt permissions
     *  see: https://github.com/aws/aws-cdk/issues/28304
     * 

> */

Author: Not necessary, removing.

new CfnOutput(this, 'ActivityArn', { value: activity.activityArn });
```

### Creating an Activity with Encryption using a Customer Managed Key

> When you provide a symmetric KMS key, all inputs/ outputs from the Step Functions Activity will be encrypted using the provided KMS key:

Author: Sorry this is incorrect only the activity inputs are encrypted. All input, output, error, and cause in the execution history including Activity-related events, are encrypted using a state machine's KMS key

});
```

## Encryption 
You can encrypt your data using a customer managed key for AWS Step Functions state machines and activities. You can configure a symmetric AWS KMS key and data key reuse period when creating or updating a State Machine or when creating an Activity. The execution history and state machine definition will be encrypted with the key applied to the State Machine. Activity inputs will be encrypted with the key applied to the Activity.

### Encrypting state machines 
You can provide a symmetric KMS key to encrypt the state machine definition and execution history:
```ts
import * as kms from 'aws-cdk-lib/aws-kms';
import * as cdk from 'aws-cdk-lib';

const kmsKey = new kms.Key(this, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  stateMachineName: 'StateMachineWithCMKEncryptionConfiguration',
  definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'Pass'))),
  stateMachineType: sfn.StateMachineType.STANDARD,
  encryptionConfiguration: new sfn.CustomerManagedEncryptionConfiguration(kmsKey, cdk.Duration.seconds(60)),
});
```

### Encrypting state machine logs in Cloud Watch Logs

> If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To optionally encrypt the logs, you can encrypt the `LogGroup` used by the state machine with a KMS key.

Author: ```suggestion
If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. If you want to encrypt the logs from the state machine using your own KMS key, you can do so by configuring the `LogGroup` associated with the state machine to use a KMS key.
```

> If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To encrypt the logs with your own KMS key you must encrypt the `LogGroup` used by the state machine with a KMS key.

Author: ```suggestion
If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To optionally encrypt the logs with your own KMS key, you can encrypt the `LogGroup` used by the state machine with a KMS key.
```

> If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To encrypt the logs with your own KMS key you must encrypt the `LogGroup` used by the state machine with a KMS key.

Author: ```suggestion
If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To optionally encrypt the logs, you can encrypt the `LogGroup` used by the state machine with a KMS key.
```

});
```

## Encryption 
You can encrypt your data using a customer-managed key for AWS Step Functions state machines and activities. You can configure a symmetric AWS KMS key and data key reuse period when creating or updating a State Machine, and when creating an Activity. The execution history and state machine definition will be encrypted with the key applied to the State Machine. Activity inputs will be encrypted with the key applied to the Activity.

### Encrypting state machines 
You can provide a symmetric KMS key to encrypt the state machine definition and execution history:

> ```

Reviewer: Set the language for syntax highlighting (repeats):

```suggestion
```ts
```

const kmsKey = new kms.Key(this, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  stateMachineName: 'StateMachineWithCMKEncryptionConfiguration',
  definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'Pass'))),
  stateMachineType: sfn.StateMachineType.STANDARD,
  kmsKey: kmsKey,
  kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(60),
});
```

> ### Creating a StateMachine with CWL Encryption using a Customer Managed Key,

Reviewer: ```suggestion
### Encrypting state machine logs in Cloud Watch Logs
```

If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To encrypt the logs with your own KMS key you must encrypt the `LogGroup` used by the state machine with a KMS key.
```
const stateMachineKmsKey = new kms.Key(this, 'StateMachine Key');
const logGroupKey = new kms.Key(this, 'LogGroup Key');

> // To enable encrypted CloudWatch log integration for a state machine, you must add the following to your log group key policy

Reviewer: This comment is somewhat misleading. The policy is just to set up an encryptied CloudWatch Logs log group. Nothing to do with the state machine really

Reviewer: Perhaps we can omit the comment entirely or trim it down:

```suggestion
// Required KMS key policy to enrypt the CloudWatch log group
```

      }),
    },
  },
}));

const logGroup = new logs.LogGroup(this, 'MyLogGroup', {
  logGroupName: '/aws/vendedlogs/states/MyLogGroup',
  encryptionKey: logGroupKey,
});

const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKWithCWLEncryption', {
  stateMachineName: 'StateMachineWithCMKWithCWLEncryption',
  definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'PassState', {
    result: sfn.Result.fromString('Hello World'),
  }))),
  stateMachineType: sfn.StateMachineType.STANDARD,
  kmsKey: stateMachineKmsKey,
  kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(300),
  logs: {
    destination: logGroup,
    level: sfn.LogLevel.ALL,
    includeExecutionData: true,
  },
});
```

### Encrypting activity inputs
When you provide a symmetric KMS key, all inputs from the Step Functions Activity will be encrypted using the provided KMS key:
```
const kmsKey = new kms.Key(this, 'Key');
const activity = new sfn.Activity(this, 'ActivityWithCMKEncryptionConfiguration', {
      activityName: 'ActivityWithCMKEncryptionConfiguration',
      kmsKey: kmsKey,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(75),

>     });

Reviewer: nit: Indentation looks off

Author: Will update.

});
```

## Encryption 
You can encrypt your data using a customer-managed key for AWS Step Functions state machines and activities. You can configure a symmetric AWS KMS key and data key reuse period when creating or updating a State Machine, and when creating an Activity. The execution history and state machine definition will be encrypted with the key applied to the State Machine. Activity inputs will be encrypted with the key applied to the Activity.

### Encrypting state machines 
You can provide a symmetric KMS key to encrypt the state machine definition and execution history:
```ts
import * as kms from 'aws-cdk-lib/aws-kms';
import * as cdk from 'aws-cdk-lib';

const kmsKey = new kms.Key(this, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  stateMachineName: 'StateMachineWithCMKEncryptionConfiguration',
  definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'Pass'))),
  stateMachineType: sfn.StateMachineType.STANDARD,
  kmsKey: kmsKey,
  kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(60),
});
```

### Encrypting state machine logs in Cloud Watch Logs

> If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To encrypt the logs with your own KMS key you must encrypt the `LogGroup` used by the state machine with a KMS key.

Reviewer: Should "customer-managed" be hyphenated? KMS and Step Functions docs don't use a hyphen

https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#customer-cmk
https://docs.aws.amazon.com/step-functions/latest/dg/encryption-at-rest.html

> If a state machine is encrypted with a customer managed key and has logging enabled, its decrypted execution history will be stored in CloudWatch Logs. To encrypt the logs with your own KMS key you must encrypt the `LogGroup` used by the state machine with a KMS key.

Author: Will omit using hyphen even though it's grammatically correct. See: https://github.com/aws/aws-cdk/pull/30959#discussion_r1736819438

  resources: ['*'],
  actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
  principals: [new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`)],
  conditions: {
    ArnEquals: {
      'kms:EncryptionContext:aws:logs:arn': cdk.Stack.of(this).formatArn({
        service: 'logs',
        resource: 'log-group',
        sep: ':',
        resourceName: '/aws/vendedlogs/states/MyLogGroup',
      }),
    },
  },
}));

const logGroup = new logs.LogGroup(this, 'MyLogGroup', {
  logGroupName: '/aws/vendedlogs/states/MyLogGroup',
  encryptionKey: logGroupKey,
});

const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKWithCWLEncryption', {
  stateMachineName: 'StateMachineWithCMKWithCWLEncryption',
  definitionBody: sfn.DefinitionBody.fromChainable(sfn.Chain.start(new sfn.Pass(this, 'PassState', {
    result: sfn.Result.fromString('Hello World'),
  }))),
  stateMachineType: sfn.StateMachineType.STANDARD,
  kmsKey: stateMachineKmsKey,
  kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(300),
  logs: {
    destination: logGroup,
    level: sfn.LogLevel.ALL,
    includeExecutionData: true,
  },

> });

Reviewer: I guess the question is whether we need a complete example of the whole thing (encrypting the log group and defining a state machine with logging and encryption). Encrypting a state machine is covered in other examples and so is setting up logging. If we are aiming for minimalism, the essential bit is encrypting the log group and setting the log group's key policy (L1005-1027).

But then again, does that code belong in the step functions README if it's a Cloud Watch Logs thing? We could omit the code and just link to the [aws_logs module REAMDE](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_logs-readme.html#encrypting-log-groups) but that seems to be missing the key policy. What do you think @paulhcsun?

> });

Reviewer 2: Hmmm with this explanation it does feel better to include encrypting the log group and setting the log group's key policy (L1005-1027). Let's keep it then but could we add in some comments to explain each part a bit more? I am not super familiar with this encryption case so it's entirely possible that this is a valid code example for the README.

> });

Author: Will update the example with more specific comments to explain the relevancy.

```

## Encryption 

> You can encrypt your data using a customer-managed key for AWS Step Functions state machines and activities. You can configure a symmetric AWS KMS key and data key reuse period when creating or updating a State Machine or when creating an Activity. The execution history and state machine definition will be encrypted with the key applied to the State Machine. Activity inputs will be encrypted with the key applied to the Activity.

Reviewer: Remove hyphen in "customer-managed" for consistency

```suggestion
You can encrypt your data using a customer managed key for AWS Step Functions state machines and activities. You can configure a symmetric AWS KMS key and data key reuse period when creating or updating a State Machine or when creating an Activity. The execution history and state machine definition will be encrypted with the key applied to the State Machine. Activity inputs will be encrypted with the key applied to the Activity.
```
